src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs

```diff
@@ -261,42 +266,136 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+                    } else {
+                        LaterUseKind::Other
                     }
                 } else {
-                    assert_eq!(location.statement_index, block.statements.len());
-                    match block.terminator().kind {
-                        TerminatorKind::Call { ref func, from_hir_call: true, .. } => {
-                            // Just point to the function, to reduce the chance
-                            // of overlapping spans.
-                            let function_span = match func {
-                                Operand::Constant(c) => c.span,
-                                Operand::Copy(Place::Local(l)) | Operand::Move(Place::Local(l)) => {
-                                    let local_decl = &self.mir.local_decls[*l];
-                                    if local_decl.name.is_none() {
-                                        local_decl.source_info.span
-                                    } else {
-                                        span
-                                    }
-                                },
-                                _ => span,
-                            };
-                            return (Call, function_span);
-                        },
-                        _ => Other,
+                    LaterUseKind::Other
+                };
+
+                (kind, span)
+            }
+        }
+    }
+
+    /// Check if a borrowed value was captured by a trait object.
+    fn was_captured_by_trait_object(&self, borrow: &BorrowData<'tcx>) -> bool {
+        // In order to check if a value was captured by a trait object, we want to look through
+        // statements after the reserve location in the current block. We expect the reserve
+        // location to be a statement assigning to a local. We follow that local in the subsequent
+        // statements, checking for an assignment of our local (or something intermediate that
+        // it was assigned into) that results in a trait object.
+        let location = borrow.reserve_location;
+        let block = &self.mir[location.block];
+        let stmt = block.statements.get(location.statement_index);
+        debug!(
+            "was_captured_by_trait_object: location={:?} block={:?} stmt={:?}",
+            location, block, stmt
+        );
+        let mut target = if let Some(&Statement {
+            kind: StatementKind::Assign(Place::Local(local), _),
+            ..
+        }) = stmt {
+            local
+        } else {
+            return false;
+        };
+
+        debug!("was_captured_by_trait_object: target={:?}", target);
+        for stmt in &block.statements[location.statement_index + 1..] {
+            debug!("was_captured_by_trait_object: stmt={:?}", stmt);
+            // Simple case where our target is assigned into another local, and we start
+            // watching that local instead.
+            if let StatementKind::Assign(
```

Comment:
Maybe we should be looking for whether the value flows into a `Cast` rvalue instead? That is, in order to assign something into a trait object, you first [*cast* it](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/enum.Rvalue.html#variant.Cast) into a trait object using [an `Unsize` cast](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/mir/enum.CastKind.html).

Reply:
Pushed a commit that fixes this.